File: 3100-3199/3120. Count the Number of Special Characters I/Solution.cpp
// Approach-1 (Using Frequency Arrays)
// T.C : O(n)
// S.C : O(1)

class Solution
{
public:
  int numberOfSpecialChars(string word)
  {

    vector<bool> lowercase(26, false);

    vector<bool> uppercase(26, false);

    for (char c : word)
    {

      if (c >= 'a' && c <= 'z')
      {

        lowercase[c - 'a'] = true;
      }
      else if (c >= 'A' && c <= 'Z')
      {

        uppercase[c - 'A'] = true;
      }
    }

    int count = 0;

    for (int i = 0; i < 26; i++)
    {

      if (lowercase[i] && uppercase[i])
      {

        count++;
      }
    }

    return count;
  }
};

/*
//Approach-2 (Using HashSet)
//T.C : O(n)
//S.C : O(n)

class Solution {
public:

    int numberOfSpecialChars(string word) {

        unordered_set<char> lower;

        unordered_set<char> upper;

        for (char ch : word) {

            if (islower(ch)) {

                lower.insert(ch);
            }
            else if (isupper(ch)) {

                upper.insert(tolower(ch));
            }
        }

        int count = 0;

        for (char ch : lower) {

            if (upper.count(ch)) {

                count++;
            }
        }

        return count;
    }
};
*/

/*
//Approach-3 (Brute Force)
//T.C : O(n^2)
//S.C : O(1)

class Solution {
public:

    int numberOfSpecialChars(string word) {

        int count = 0;

        vector<bool> arr(26, false);

        for (int i = 0; i < word.length(); i++) {

            char ch = word[i];

            if (isupper(ch)) {

                continue;
            }
            else {

                char upper = toupper(ch);

                for (int j = 0; j < word.length(); j++) {

                    if (word[j] == upper && !arr[ch - 'a']) {

                        count++;

                        arr[ch - 'a'] = true;
                    }
                }
            }
        }

        return count;
    }
};
*/

File: 3100-3199/3120. Count the Number of Special Characters I/Solution.java
//Approach-1 (Using Frequency Arrays)
//T.C : O(n)
//S.C : O(1)

class Solution {

    public int numberOfSpecialChars(String word) {

        boolean[] lowercase = new boolean[26];

        boolean[] uppercase = new boolean[26];

        for (char c : word.toCharArray()) {

            if (c >= 'a' && c <= 'z') {

                lowercase[c - 'a'] = true;
            }
            else if (c >= 'A' && c <= 'Z') {

                uppercase[c - 'A'] = true;
            }
        }

        int count = 0;

        for (int i = 0; i < 26; i++) {

            if (lowercase[i] && uppercase[i]) {

                count++;
            }
        }

        return count;
    }
}


/*
//Approach-2 (Using HashSet)
//T.C : O(n)
//S.C : O(n)

class Solution {

    public int numberOfSpecialChars(String word) {

        Set<Character> lower = new HashSet<>();

        Set<Character> upper = new HashSet<>();

        for (char ch : word.toCharArray()) {

            if (Character.isLowerCase(ch)) {

                lower.add(ch);
            }
            else if (Character.isUpperCase(ch)) {

                upper.add(Character.toLowerCase(ch));
            }
        }

        int count = 0;

        for (char ch : lower) {

            if (upper.contains(ch)) {

                count++;
            }
        }

        return count;
    }
}
*/


/*
//Approach-3 (Brute Force)
//T.C : O(n^2)
//S.C : O(1)

class Solution {

    public int numberOfSpecialChars(String word) {

        int count = 0;

        boolean[] arr = new boolean[26];

        Arrays.fill(arr, false);

        for (int i = 0; i < word.length(); i++) {

            char ch = word.charAt(i);

            if (Character.isUpperCase(ch)) {

                continue;
            }
            else {

                char upper = Character.toUpperCase(ch);

                for (int j = 0; j < word.length(); j++) {

                    if (word.charAt(j) == upper && !arr[ch - 'a']) {

                        count++;

                        arr[ch - 'a'] = true;
                    }
                }
            }
        }

        return count;
    }
}
*/